examples/imagenet/imagenet_lib.py

```diff
@@ -191,58 +159,78 @@ def _prepare(x):
   return jax.tree_map(_prepare, xs)
 
 
-def create_input_iter(batch_size, image_size, dtype, train, cache):
-  ds = input_pipeline.load_split(
-      batch_size, image_size=image_size, dtype=dtype, train=train, cache=cache)
-  it = map(prepare_tf_data, ds)
+def _create_input_iter(dataset_builder, batch_size, image_size, dtype, train,
+                       cache):
+  ds = input_pipeline.create_split(
+      dataset_builder, batch_size, image_size=image_size, dtype=dtype,
+      train=train, cache=cache)
+  it = map(_prepare_tf_data, ds)
   it = jax_utils.prefetch_to_device(it, 2)
   return it
 
 
 # flax.struct.dataclass enables instances of this class to be passed into jax
 # transformations like tree_map and pmap.
 @flax.struct.dataclass
-class TrainState:
+class _TrainState:
   step: int
   optimizer: optim.Optimizer
   model_state: nn.Collection
   dynamic_scale: optim.DynamicScale
 
 
-def restore_checkpoint(state):
-  return checkpoints.restore_checkpoint(FLAGS.model_dir, state)
+def _restore_checkpoint(model_dir, state):
+  return checkpoints.restore_checkpoint(model_dir, state)
 
 
-def save_checkpoint(state):
+def _save_checkpoint(model_dir, state):
   if jax.host_id() == 0:
     # get train state from the first replica
     state = jax.device_get(jax.tree_map(lambda x: x[0], state))
     step = int(state.step)
-    checkpoints.save_checkpoint(FLAGS.model_dir, state, step, keep=3)
+    checkpoints.save_checkpoint(model_dir, state, step, keep=3)
 
 
-def sync_batch_stats(state):
+def _sync_batch_stats(state):
   """Sync the batch statistics across replicas."""
   avg = jax.pmap(lambda x: lax.pmean(x, 'x'), 'x')
   return state.replace(model_state=avg(state.model_state))
 
 
-def main(argv):
-  if len(argv) > 1:
-    raise app.UsageError('Too many command-line arguments.')
+def train_and_evaluate(model_dir: str, batch_size: int, num_epochs: int,
+                       learning_rate: float, momentum: float, cache: bool,
+                       half_precision: bool,
+                       num_train_and_eval_steps: int = -1,
+                       disable_checkpointing: bool = False):
+  """Runs model training and evaluation loop.
+
+  Args:
+    model_dir: Directory where the checkpoints and tensorboard summaries
+      should be written to.
+    batch_size: Batch size of the input.
+    num_epochs: Number of epochs to cycle through before stopping.
+    learning_rate: Learning rate for the momentum optimizer.
+    momentum: Momentum value for the momentum optimizer.
+    cache: Determines whether the dataset should be cached.
+    half_precision: Determines whether bfloat16/float16 should be used
+      instead of float32.
+    num_train_and_eval_steps: Number of steps for training and eval.
```

Comment:
Hmm... Do we need this? It's a bit off to see an argument designed just for a test. Are there any other ways to achieve this?

Reply:
(I wrote another comment on this below where you call this function from your test)

Reply:
Although specifying the number of examples for train and test/eval in tfds.testing.mockdata, the base info such as num_examples in a split (train/eval) is same as the original dataset. It is documented as WARNING in [documentation](https://www.tensorflow.org/datasets/api_docs/python/tfds/testing/mock_data).

Without this flag, we would do a lot more than a single iteration :/ I couldn't think of something better.. 

We can discuss further in the below comment if refactoring is the way to go.

Reply:
The WARNING in the documentation says that `info.split['train'].num_examples` doesn't match `len(list(ds_train))`. So how about you use `len(list(ds_train))` below instead of `.num_examples` to determine the size of the dataset? In that case I think you won't need this argument.